aten/src/ATen/native/TensorIteratorReduce.cpp

```diff
@@ -35,36 +36,32 @@ static bool use_two_pass_reduction(TensorIteratorBase& iter) {
 }
 
 static void two_pass_reduction(TensorIteratorBase& iter, loop2d_t loop) {
-  int max_threads = at::get_num_threads();
+  const int max_threads = at::get_num_threads();
 
   auto dst = iter.output(0);
-  auto buffer_shape = DimVector(dst.sizes());
-  buffer_shape.insert(buffer_shape.begin(), max_threads);
+  auto unsqueezed = dst.unsqueeze(0);
+  auto buffer_shape = DimVector(unsqueezed.sizes());
+  buffer_shape[0] = max_threads;
   auto buffer = at::empty(buffer_shape, dst.options());
+  // Fill with the identity
+  buffer.copy_(unsqueezed);
 
-  // NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays,modernize-avoid-c-arrays)
-  std::unique_ptr<bool[]> written(new bool[max_threads]);
-  std::fill(written.get(), written.get() + max_threads, false);
+  auto buffer_stride = buffer.strides()[0] * buffer.element_size();
+  auto buffer_0 = buffer[0];
+  auto first_reduce = TensorIterator::reduce_op(buffer_0, iter.input(0));
 
   at::parallel_for(0, iter.numel(), internal::GRAIN_SIZE, [&](int64_t begin, int64_t end) {
-    int thread_num = at::get_thread_num();
-    // NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays,modernize-avoid-c-arrays)
-    written[thread_num] = true;
-    auto slice = buffer[thread_num];
-    slice.copy_(dst);
-
-    auto sub_iter = TensorIterator::reduce_op(slice, iter.input(0));
-    sub_iter.serial_for_each(loop, {begin, end});
-  });
+    const auto thread_num = at::get_thread_num();
+    auto shape = first_reduce.shape();
+    auto strides = first_reduce.get_strides();
 
-  // fill any unwritten slices of the buffer with the identity
-  for (int thread_num = 0; thread_num < max_threads; thread_num++) {
-    if (!written[thread_num]) {
-      buffer[thread_num].copy_(dst);
-    }
-  }
+    // Bump output ptr so each thread has its own ouput slice
+    auto base_ptrs = first_reduce.get_base_ptrs();
```

Comment:
what you are doing here can probably be achieved by `unsafe_replace_operand` method in TensorIterator?

Reply:
TBH I didn't know about `unsafe_replace_operand`. However, since it modifies the iterator, I would need to take an extra copy of the `TensorIterator` on each thread. In terms of callgrind, with this PR I see for a 1-element sum() and inclusive counts:
```
  433753  ???:at::two_pass_reduction(at::T ... har**, long const*, long, long)>)
```
and using `unsafe_replace_operand` it's
```
  452233  ???:at::two_pass_reduction(at::T ... har**, long const*, long, long)>)
```

So on the order if 5% overhead. That said, this is only used for large reductions anyway so perhaps that's okay.

Reply:
You are changing `TensorIteratorBase::serial_for_each(loop2d_t loop, Range range)` which is used everywhere, so can you please benchmark a regular unary and binary pointwise operation?